packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -2514,7 +2519,8 @@ class _RouteEntry extends RouteTransitionRecord {
       'This route cannot be marked for pop. Either a decision has already been '
       'made or it does not require an explicit decision on how to transition out.'
     );
-    pop<dynamic>(result);
+    if (isPresent)
```

Comment:
This is odd, if the route is no longer present, the transition delegate shouldn't be allowed to make a decision about how it should exit (especially if that decision just gets ignored). To say it with the assert above: "[The route] does not require an explicit decision on how to transition out" because, well, a decision has already been made... The only thing the transition delegate is allowed to do with them is to re-arrenge them.

Reply:
We may need to extend the transition delegate to accommodate this...

Reply:
yes, either this or we break the transition delegate api. I honestly don't know which is better, It is a corner case that should not happen too often. I wonder if it will be even more complicated to handle this in transition delegate given how complicated it already was.

Reply:
Breaking the transition delegate API should be fine. The API is so new and shouldn't be in much use it.
In the transition delegate could these pages just be part of the locationToExitingPageRoute list and there entry is marked somehow to indicate that this are already in the process of exiting?

Reply:
There is another bug, if a new page matches the a popping route, that route will be updated but still popped.

It can easily be reproduced If user rebuilds the navigator without the removing the page from the pages list in onPopPage callback.

Reply:
That does sound like a bug. Since our documentation says:

"The next time the [pages] list is updated, if the [Page] corresponding to this [Route] is still present, it will be interpreted as a new route to display."

we should in updatePage ignore pages that are on their way out and treat that page as a new one.

Reply:
actually this seems to be taken cared of already where we check the ```currentState.index > _RouteLifecycle.idle.index``` in canUpdateFrom